File: cli/src/commands/refresh-metadata.ts
import type { PhotoAsset } from "@attic/shared";
import { buildMetadataJson, metadataKey } from "@attic/shared";
import type { Manifest } from "../manifest/manifest.ts";
import type { S3Provider } from "../storage/s3-client.ts";
import { formatBytes } from "../format.ts";
import { withRetry } from "../retry.ts";

export interface RefreshMetadataOptions {
  /** Maximum concurrent uploads. */
  concurrency: number;
  /** Show what would be uploaded without uploading. */
  dryRun: boolean;
}

const DEFAULT_OPTIONS: RefreshMetadataOptions = {
  concurrency: 20,
  dryRun: false,
};

export interface RefreshMetadataReport {
  updated: number;
  skipped: number;
  failed: number;
  totalBytes: number;
  errors: Array<{ uuid: string; message: string }>;
}

/**
 * Re-upload metadata JSON for already-backed-up assets.
 * Original files and manifest are left untouched.
 */
export async function refreshMetadata(
  assets: PhotoAsset[],
  manifest: Manifest,
  s3: S3Provider,
  opts: Partial<RefreshMetadataOptions> = {},
): Promise<RefreshMetadataReport> {
  const options = { ...DEFAULT_OPTIONS, ...opts };
  options.concurrency = Math.max(1, options.concurrency);

  // Only refresh assets that are in the manifest
  const assetByUuid = new Map<string, PhotoAsset>();
  for (const a of assets) {
    assetByUuid.set(a.uuid, a);
  }

  const toRefresh: Array<
    { asset: PhotoAsset; s3Key: string; checksum: string; backedUpAt: string }
  > = [];
  for (const [uuid, entry] of Object.entries(manifest.entries)) {
    const asset = assetByUuid.get(uuid);
    if (asset) {
      toRefresh.push({
        asset,
        s3Key: entry.s3Key,
        checksum: entry.checksum,
        backedUpAt: entry.backedUpAt,
      });
    }
  }

  console.log(`\n  Attic — Refresh Metadata`);
  console.log(`  ════════════════════════\n`);
  console.log(
    `  Backed-up assets in DB:  ${toRefresh.length.toLocaleString()}`,
  );
  if (options.dryRun) console.log(`  Mode:                    DRY RUN`);
  console.log();

  if (toRefresh.length === 0) {
    console.log("  Nothing to refresh — no backed-up assets found in DB.\n");
    return { updated: 0, skipped: 0, failed: 0, totalBytes: 0, errors: [] };
  }

  if (options.dryRun) {
    return {
      updated: 0,
      skipped: toRefresh.length,
      failed: 0,
      totalBytes: 0,
      errors: [],
    };
  }

  const report: RefreshMetadataReport = {
    updated: 0,
    skipped: 0,
    failed: 0,
    totalBytes: 0,
    errors: [],
  };

  // Process with bounded concurrency using an index counter (O(1) per item).
  // Mutations to `report` are safe: Deno is single-threaded, and all
  // increments happen synchronously between await points.
  let cursor = 0;
  const workers = Array.from(
    { length: Math.min(options.concurrency, toRefresh.length) },
    async () => {
      while (cursor < toRefresh.length) {
        const item = toRefresh[cursor++];
        try {
          const meta = buildMetadataJson(
            item.asset,
            item.s3Key,
            item.checksum,
            item.backedUpAt,
          );
          const data = new TextEncoder().encode(
            JSON.stringify(meta, null, 2),
          );
          await withRetry(() =>
            s3.putObject(
              metadataKey(item.asset.uuid),
              data,
              "application/json",
            )
          );
          report.updated++;
          report.totalBytes += data.byteLength;
        } catch (error: unknown) {
          const msg = error instanceof Error ? error.message : "Unknown error";
          report.errors.push({ uuid: item.asset.uuid, message: msg });
          report.failed++;
        }

        const done = report.updated + report.failed;
        if (done % 50 === 0 || done === toRefresh.length) {
          const pct = ((done / toRefresh.length) * 100).toFixed(1);
          console.log(
            `    Progress: ${done}/${toRefresh.length} (${pct}%)  ` +
              `Uploaded: ${formatBytes(report.totalBytes)}`,
          );
        }
      }
    },
  );

  await Promise.all(workers);

  console.log(`\n  ── Complete ──`);
  console.log(`  Updated:   ${report.updated.toLocaleString()}`);
  console.log(`  Failed:    ${report.failed.toLocaleString()}`);
  console.log(`  Total:     ${formatBytes(report.totalBytes)}\n`);

  return report;
}

File: Sources/AtticCore/RefreshMetadata.swift
import Foundation
import LadderKit

/// Options controlling refresh-metadata behavior.
public struct RefreshMetadataOptions: Sendable {
    public var concurrency: Int
    public var dryRun: Bool

    public init(concurrency: Int = 20, dryRun: Bool = false) {
        self.concurrency = concurrency
        self.dryRun = dryRun
    }
}

/// Result of a refresh-metadata run.
public struct RefreshMetadataReport: Sendable {
    public var updated: Int = 0
    public var skipped: Int = 0
    public var failed: Int = 0
    public var totalBytes: Int = 0
    public var errors: [(uuid: String, message: String)] = []
}

/// Progress events emitted during metadata refresh.
public protocol RefreshMetadataProgressDelegate: Sendable {
    func refreshStarted(total: Int)
    func assetRefreshed(uuid: String, filename: String)
    func assetFailed(uuid: String, message: String)
    func refreshCompleted(report: RefreshMetadataReport)
}

/// No-op refresh progress delegate.
public struct NullRefreshMetadataProgressDelegate: RefreshMetadataProgressDelegate {
    public init() {}
    public func refreshStarted(total: Int) {}
    public func assetRefreshed(uuid: String, filename: String) {}
    public func assetFailed(uuid: String, message: String) {}
    public func refreshCompleted(report: RefreshMetadataReport) {}
}

/// Re-generate and upload metadata JSON for all backed-up assets.
///
/// This is useful when the metadata schema changes or new fields are added.
/// Only processes assets that exist in both the manifest and the asset list.
public func runRefreshMetadata(
    assets: [AssetInfo],
    manifest: Manifest,
    s3: any S3Providing,
    options: RefreshMetadataOptions = RefreshMetadataOptions(),
    progress: any RefreshMetadataProgressDelegate = NullRefreshMetadataProgressDelegate(),
) async throws -> RefreshMetadataReport {
    // Only refresh assets that are in the manifest
    let backedUp = assets.filter { manifest.isBackedUp($0.uuid) }

    guard !backedUp.isEmpty else {
        let report = RefreshMetadataReport()
        progress.refreshCompleted(report: report)
        return report
    }

    progress.refreshStarted(total: backedUp.count)

    if options.dryRun {
        var report = RefreshMetadataReport()
        report.skipped = backedUp.count
        progress.refreshCompleted(report: report)
        return report
    }

    let report = RefreshReport()

    await withTaskGroup(of: Void.self) { group in
        var cursor = 0

        for _ in 0 ..< min(options.concurrency, backedUp.count) {
            let asset = backedUp[cursor]
            cursor += 1
            group.addTask {
                await refreshSingle(
                    asset: asset,
                    manifest: manifest,
                    s3: s3,
                    report: report,
                    progress: progress,
                )
            }
        }

        for await _ in group {
            if cursor < backedUp.count {
                let asset = backedUp[cursor]
                cursor += 1
                group.addTask {
                    await refreshSingle(
                        asset: asset,
                        manifest: manifest,
                        s3: s3,
                        report: report,
                        progress: progress,
                    )
                }
            }
        }
    }

    let finalReport = await report.snapshot()
    progress.refreshCompleted(report: finalReport)
    return finalReport
}

// MARK: - Internals

private actor RefreshReport {
    var updated = 0
    var failed = 0
    var totalBytes = 0
    var errors: [(uuid: String, message: String)] = []

    func markUpdated(bytes: Int) {
        updated += 1
        totalBytes += bytes
    }

    func markFailed(_ uuid: String, _ message: String) {
        failed += 1
        if errors.count < maxReportErrors {
            errors.append((uuid: uuid, message: message))
        }
    }

    func snapshot() -> RefreshMetadataReport {
        RefreshMetadataReport(
            updated: updated, skipped: 0, failed: failed,
            totalBytes: totalBytes, errors: errors,
        )
    }
}

private func refreshSingle(
    asset: AssetInfo,
    manifest: Manifest,
    s3: any S3Providing,
    report: RefreshReport,
    progress: any RefreshMetadataProgressDelegate,
) async {
    guard let entry = manifest.entries[asset.uuid] else { return }

    do {
        let meta = buildMetadataJSON(
            asset: asset,
            s3Key: entry.s3Key,
            checksum: entry.checksum,
            backedUpAt: entry.backedUpAt,
        )
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        let data = try encoder.encode(meta)
        let metaKey = try S3Paths.metadataKey(uuid: asset.uuid)

        try await withRetry {
            try await s3.putObject(key: metaKey, body: data, contentType: "application/json")
        }

        await report.markUpdated(bytes: data.count)
        let filename = asset.originalFilename ?? asset.uuid
        progress.assetRefreshed(uuid: asset.uuid, filename: filename)
    } catch {
        let msg = String(describing: error)
        await report.markFailed(asset.uuid, msg)
        progress.assetFailed(uuid: asset.uuid, message: msg)
    }
}
